cosmos/io.py

```diff
@@ -0,0 +1,183 @@
+    azure_conn_id = kwargs.get("azure_conn_id", WasbHook.default_conn_name)
+    container_name = kwargs["container_name"]
+    hook = WasbHook(wasb_conn_id=azure_conn_id)
+
+    # Iterate over the files in the target dir and upload them to WASB container
+    for dirpath, _, filenames in os.walk(target_dir):
+        for filename in filenames:
+            blob_name = (
+                f"{kwargs['dag'].dag_id}"
+                f"/{kwargs['run_id']}"
+                f"/{kwargs['task_instance'].task_id}"
+                f"/{kwargs['task_instance']._try_number}"
+                f"{dirpath.split(project_dir)[-1]}/{filename}"
+            )
+            hook.load_file(
+                file_path=f"{dirpath}/{filename}",
+                container_name=container_name,
+                blob_name=blob_name,
+                overwrite=True,
+            )
+
+
+def _configure_remote_target_path() -> tuple[Path, str] | tuple[None, None]:
+    """Configure the remote target path if it is provided."""
+    from airflow.version import version as airflow_version
+
+    if not remote_target_path:
+        return None, None
+
+    _configured_target_path = None
+
+    target_path_str = str(remote_target_path)
+
+    remote_conn_id = remote_target_path_conn_id
+    if not remote_conn_id:
+        target_path_schema = urlparse(target_path_str).scheme
+        remote_conn_id = FILE_SCHEME_AIRFLOW_DEFAULT_CONN_ID_MAP.get(target_path_schema, None)  # type: ignore[assignment]
+    if remote_conn_id is None:
+        return None, None
+
+    if not settings.AIRFLOW_IO_AVAILABLE:
+        raise CosmosValueError(
+            f"You're trying to specify remote target path {target_path_str}, but the required "
+            f"Object Storage feature is unavailable in Airflow version {airflow_version}. Please upgrade to "
+            "Airflow 2.8 or later."
+        )
+
+    from airflow.io.path import ObjectStoragePath
+
+    _configured_target_path = ObjectStoragePath(target_path_str, conn_id=remote_conn_id)
+
+    if not _configured_target_path.exists():  # type: ignore[no-untyped-call]
+        _configured_target_path.mkdir(parents=True, exist_ok=True)
+
+    return _configured_target_path, remote_conn_id
+
+
+def _construct_dest_file_path(
+    dest_target_dir: Path,
+    file_path: str,
+    source_target_dir: Path,
+    **kwargs: Any,
+) -> str:
+    """
+    Construct the destination path for the artifact files to be uploaded to the remote store.
+    """
+    dest_target_dir_str = str(dest_target_dir).rstrip("/")
+
+    task_run_identifier = (
+        f"{kwargs['dag'].dag_id}"
+        f"/{kwargs['run_id']}"
+        f"/{kwargs['task_instance'].task_id}"
+        f"/{kwargs['task_instance']._try_number}"
+    )
+    rel_path = os.path.relpath(file_path, source_target_dir).lstrip("/")
+
+    return f"{dest_target_dir_str}/{task_run_identifier}/target/{rel_path}"
+
+
+def upload_artifacts_to_cloud_storage(project_dir: str, **kwargs: Any) -> None:
```

Comment:
How does this relate to the other upload methods? Is this generic, regardless of the target object store?

Reply:
do you mean to ask if this applies to all of S3, GCP, Azure? If that's the question, then yes, this method applies to all of them. Airflow Object Stores figures it out to upload them to corresponding object stores.